File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {

    // subham: https://www.scaler.com/help_requests/694867/?ref=join-call-now    
    public int search(int[] nums, int target) {
        int n = nums.length;
        
        int pivot = findPivot(nums);

        int idx1 = Arrays.binarySearch(nums, 0, pivot, target);
        int idx2 = Arrays.binarySearch(nums, pivot, n, target);

        // for(int idx : new int[]{idx1, idx2}){
        //     if(idx < 0){
        //         idx = -idx - 1;
        //     }
        //     if (idx < n && nums[idx] == target) 
        //         return idx;
        // }
        
        if (idx1 < 0)
            idx1 = -idx1 - 1;   // -1 because index can be at max -1
        
        if (idx2 < 0)
            idx2 = -idx2 - 1;
        
        if (idx1 < n && nums [idx1] == target)
            return idx1;
        else if (idx2 < n && nums [idx2] == target)
            return idx2;
        
        return -1;
    }
    
    private int findPivot(int[] nums) 
    {
        int left = 0;
        int right = nums.length - 1;
        
        while (left < right) {
            int mid = left + (right - left) / 2;
            int minVal = Math.min(nums[mid], Math.min(nums[left], nums[right]));
            if (minVal == nums[right])
                left = mid + 1;
            else
                right = mid;
        }

        return left;
    }
}

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        return self.helper( nums, target)
    
    
    def helper(self, A, target):

        if len(A) == 1:
            if A[0] == target:
                return 0
            return -1

        pivotIndex = self.getPivotIndex(A)
        
        # print(pivotIndex)
        
        if target >= A[0] and target <= A[pivotIndex]:
            return self.binarySearch(A, target, 0, pivotIndex)
        
        if pivotIndex + 1 < len(A) and target >= A[pivotIndex + 1] and target <= A[len(A) - 1]:
            return self.binarySearch(A, target, pivotIndex + 1, len(A) - 1)
        
        return -1
    
    def getPivotIndex(self, A):
            left, right, n = 0, len(A) - 1, len(A)
            
            while left <= right:
                
                mid = left + (right - left) // 2
                
                if mid + 1 < n and A[mid] > A[mid + 1]:
                    return mid
                
                if mid - 1 >= 0 and A[mid - 1] > A[mid]:
                    return mid - 1
                
                if A[left] >= A[mid]:
                    right = mid - 1
                else:
                    left = mid + 1
                    
            return 0
        
    def binarySearch(self, A, target, left, right):
        while left <= right:
            mid = left + (right - left) // 2

            if A[mid] == target:
                return mid

            if A[mid] > target:
                right = mid - 1
            else:
                left = mid + 1

        return -1